### ISSUE TITLE
Any & ~Any used through alias incorrectly simplifies to Never

### ISSUE THREAD

**Issue by JelleZijlstra** [astral-sh/ty#4062]
### Summary

```python
from ty_extensions import Intersection, Not
from typing import Any

type D[T] = Intersection[T, Not[T]]

def f(x: D[Any]):
    reveal_type(x) # Never
```
But when you use `Any & ~Any` directly, it remains `Any`, which is the answer I expect.

Is this because ty does something clever with type aliases that makes it so both Anys have to stay the same? 

### Version

0.0.62

**Comment by JelleZijlstra** [comment]
It doesn't appear that ty generally assumes that all arguments to a type alias must stay the "same" Any, because it allows this:

```python
from ty_extensions import Intersection, Not
from typing import Any

type Pair[T] = tuple[T, T]

def f(x: list[Pair[Any]]):
    y: list[tuple[int, str]] = x
```

**Comment by sharkdp** [comment]
I think we eagerly simplify `T & ~T` to `Never`, which seems correct to me. Observe, for example:

```py
def g[T](x: D[T]):
    reveal_type(x)  # also Never
```

> Is this because ty does something clever with type aliases that makes it so both Anys have to stay the same?

So I think it's not the type alias, but rather the type variable `T` that keeps the identity between the positive and negative part, which is needed to show that `T & ~T = Never`. This is fundamentally different from `Any & ~Any` where the two `Any`'s could materialize to different types.

> It doesn't appear that ty generally assumes that all arguments to a type alias must stay the "same" Any, because it allows this:

Yes, here, we explicitly specialize `Pair` with `Any`, which creates the type `tuple[Any, Any]`, which (unfortunately?) loses that "the two `Any`'s have to be the same" property.



So in summary, I think ty's behavior here is correct and desirable?

The fundamental problem that we encounter here (yet again) is that the syntactical description of gradual types is very much insufficient to describe the full universe of gradual types. It's not just the problem with unions/intersections of materializations (https://github.com/python/typing/issues/2027), which is insufficiently described by the syntactical form of gradual types, it's also the fact that we would like to be able to "label" occurrences of `Any` in order to be able to say: these two `Any`'s need to materialize to the same type. If we had that, `D[Any]` would be the type `Any₁ & ~Any₁`, which is equal to `Never`, and what we typically just write as `Any & ~Any` would be the gradual type `Any₁ & ~Any₂` which is equal to `Any₁`.

And in your second example, we would like `Pair[Any]` to be `tuple[Any₁, Any₁]`, instead of `tuple[Any₁, Any₂]` (which is what ty currently creates).

Or in other words, type aliases would ideally give us the flexibility to express these gradual types with "labeled `Any`'s", but ty (and other type checkers) do not treat them consistently yet.

**Comment by carljm** [comment]
Fully agree with the above, with one minor quibble:

> what we typically just write as `Any & ~Any` would be the gradual type `Any₁ & ~Any₂` which is equal to `Any₁`.

I think if we tracked Any identities, `Any₁ & ~Any₂` could not correctly simplify to `Any₁`, but would either need to remain unsimplified (not lossy), or would have to become a different `Any₃` (lossy).

**Comment by JelleZijlstra** [comment]
OK, simplifying the TypeVars on the theory that they represent a single type is reasonable and consistent with ty's behavior. Thanks for discussing this!

**Comment by sharkdp** [comment]
> Fully agree with the above, with one minor quibble:
> 
> > what we typically just write as `Any & ~Any` would be the gradual type `Any₁ & ~Any₂` which is equal to `Any₁`.
> 
> I think if we tracked Any identities, `Any₁ & ~Any₂` could not correctly simplify to `Any₁`, but would either need to remain unsimplified (not lossy), or would have to become a different `Any₃` (lossy).

Just for completeness, we discussed this and there was just a misunderstanding in how we thought that notation would work. A hopefully less confusing way to put it:

Given
```py
type G1[X, Y] = X & ~Y
type G2[X] = X & ~X
```
we have `G1[Any, Any] = Any` and `G2[Any] = Never`.

Above, I was referring to `G1[Any, Any]` as `Any₁ & ~Any₂`, and to `G2[Any]` as `Any₁ & ~Any₁`. In that interpretation, `Any₁` is also equivalent to `Any₃` or to `Any` with any other label, because the label is just scoped to that gradual type, not tracked further in the scope or similar.
